### PR TITLE
fix(op): resolve client before RedirectToLogin for custom AuthorizeValidator

### LINKED ISSUE #909: [Bug]: op.Authorize nil-pointer panic in RedirectToLogin when a custom AuthorizeValidator is used
### Preflight Checklist

- I could not find an existing issue for this bug (searched `AuthorizeValidator`, `RedirectToLogin`, `Authorize` + `nil`/`panic`; #315/#358 are a different, already-fixed parse-error path).

### Version

v3.47.5 (also present on `main`).

### Describe the bug

`op.Authorize` panics with a nil-pointer dereference in `RedirectToLogin` when the `Authorizer` also implements the `op.AuthorizeValidator` interface (i.e. a custom `ValidateAuthRequest`).

In `pkg/op/auth_request.go`, `Authorize` declares `var client Client` and populates it **only inside the default `validation` closure**:

```go
var client Client
validation := func(ctx context.Context, authReq *oidc.AuthRequest, storage Storage, verifier *IDTokenHintVerifier) (sub string, err error) {
    client, err = authorizer.Storage().GetClientByClientID(ctx, authReq.ClientID) // sets outer `client`
    ...
}
if validator, ok := authorizer.(AuthorizeValidator); ok {
    validation = validator.ValidateAuthRequest // replaces the closure -> outer `client` never assigned
}
userID, err := validation(...)
...
RedirectToLogin(req.GetID(), client, w, r) // client is nil when a custom validator was used
```

When a custom `AuthorizeValidator` is supplied, `validation` is replaced, so the outer `client` stays `nil`, and `RedirectToLogin` -> `client.LoginURL(authReqID)` dereferences a nil `Client` and panics.

This is unfortunate because `AuthorizeValidator` is documented as the intended seam to "customize validation", yet using it crashes the standard login-redirect path for any well-formed auth request.

### Steps to Reproduce

1. Build an `Authorizer` that also implements `op.AuthorizeValidator` with a `ValidateAuthRequest` that returns a valid `userID` (as a custom validator naturally would, without re-assigning the library's internal `client`).
2. Send a well-formed `/authorize` request (valid `client_id`, `redirect_uri`, `response_type=code`, `scope=openid`) whose flow requires a login redirect.
3. `op.Authorize` panics: `runtime error: invalid memory address or nil pointer dereference` at `RedirectToLogin` (`client.LoginURL`).

### Expected behavior

The `AuthorizeValidator` seam should be usable without crashing: `client` should be resolved before `RedirectToLogin`.

### Proposed fix

Ensure `client` is populated before the redirect when a custom validator left it unset. Minimal, and it passes the existing `pkg/op` test suite:

```go
// after the validation() error check, before CreateAuthRequest / RedirectToLogin
if client == nil {
    client, err = authorizer.Storage().GetClientByClientID(ctx, authReq.ClientID)
    if err != nil {
        AuthRequestError(w, r, authReq, oidc.DefaultToServerError(err, "unable to retrieve client by id"), authorizer)
        return
    }
}
```

Happy to open a PR (with a regression test using a mock `Authorizer` that implements `AuthorizeValidator`) if the approach looks right to you.

### Relevant Configuration / Additional Context

Discovered while building an IdP that uses a custom `AuthorizeValidator` to make PKCE (S256) mandatory for all client types.

### PR DESCRIPTION
**Pull request by abuobaida**
### Definition of Ready

- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the PR description
- [x] PR is linked to the corresponding user story

## Which Problems Are Solved

Fixes #909.

`op.Authorize` panics with a nil-pointer dereference in `RedirectToLogin` when the
`Authorizer` implements the `op.AuthorizeValidator` interface. `Authorize` populates
its local `client` only inside the *default* validation closure; supplying a custom
`AuthorizeValidator` replaces that closure, so `client` stays `nil` and
`RedirectToLogin` (`client.LoginURL`) panics for any well-formed auth request.

## How the Problems Are Solved

Before the login redirect, if `client` is still `nil` (a custom validator was used and
did not assign it), resolve it via `Storage().GetClientByClientID`, mapping a lookup
error to the standard `AuthRequestError` path.

## Additional Context

- The `AuthorizeValidator` interface is documented as the seam to customize validation,
  but using it currently crashes the standard login-redirect path.
- The change passes the existing `pkg/op` test suite unchanged.
- Happy to add a regression test in your preferred style — the path needs an
  `Authorizer` that also implements `AuthorizeValidator`, and `op.AuthRequest` has no
  mock today, so I wanted to confirm the approach/test conventions with you first.

### CODE REVIEW

**Review by wim07101993 — CHANGES_REQUESTED** [review]
Hi, thank you for the contribution. Could you perhaps also add a test which triggers the nil-pointer panic?

**Comment by abuobaida** [comment]
Thanks for the review @wim07101993 — added `TestAuthorizeCustomValidator` in 34bc30f. It drives `op.Authorize` with an `Authorizer` that also implements `AuthorizeValidator` whose validation succeeds without resolving the client. Verified against the pre-fix code it panics with the exact #909 nil-pointer dereference in `RedirectToLogin`; with the fix it asserts the login redirect.

File: pkg/op/auth_request.go [hunk 1]
```diff
@@ -128,6 +128,17 @@ func Authorize(w http.ResponseWriter, r *http.Request, authorizer Authorizer) {
...
+	// A custom AuthorizeValidator (see the AuthorizeValidator interface) replaces the
+	// default validation closure that assigns client above, so client may still be nil
+	// here. RedirectToLogin dereferences client (client.LoginURL), so fetch it if a
+	// custom validator left it unset, to avoid a nil pointer panic.
+	if client == nil {
+		client, err = authorizer.Storage().GetClientByClientID(ctx, authReq.ClientID)
+		if err != nil {
+			AuthRequestError(w, r, authReq, oidc.DefaultToServerError(err, "unable to retrieve client by id"), authorizer)
+			return
+		}
```
[the remark this answers is not in the document]
Comment (wim07101993) [reply pkg/op/auth_request.go:142]: could you fix this one, then all seems to be ok

**Review by wim07101993 — CHANGES_REQUESTED** [review]

File: pkg/op/auth_request.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (abuobaida) [reply pkg/op/auth_request.go:142]: Fixed in 768da54 — the fallback lookup failure now goes through `oidc.ErrInvalidRequestRedirectURI()` (same construction as the default validation closure), so `AuthRequestError` renders the error directly instead of redirecting to the unvalidated `redirect_uri`. Added `TestAuthorizeCustomValidatorClientLookupError` asserting a direct 400 with no `Location` header even though the request carries a `redirect_uri`. Thanks @wim07101993!

**Review by wim07101993 — APPROVED** [review]
